metric_learn/scml.py

```diff
@@ -0,0 +1,647 @@
+
+    if self.n_basis is None:
+      # TODO: Get a good default n_basis directive
+      n_basis = min(20*n_features, X.shape[0]*2*num_eig)
+      warnings.warn('As no value for `n_basis` was selected, the number of '
+                    'basis will be set to n_basis= %d' % n_basis)
+
+    elif isinstance(self.n_basis, int):
+      n_basis = self.n_basis
+    else:
+      raise ValueError("n_basis should be an integer, instead it is of type %s"
+                       % type(self.n_basis))
+
+    if n_basis <= n_class:
+      raise ValueError("The number of basis should be greater than the"
+                       " number of classes")
+    elif n_basis >= X.shape[0]*2*num_eig:
+      raise ValueError("The selected number of basis needs a greater number of"
+                       " clusters than the number of available samples")
+
+    # Number of clusters needed for 2 scales given the number of basis
+    # yielded by every LDA
+    n_clusters = int(np.ceil(n_basis/(2 * num_eig)))
+
+    # TODO: maybe give acces to Kmeans jobs for faster computation?
+    kmeans = KMeans(n_clusters=n_clusters, random_state=self.random_state,
+                    algorithm='elkan').fit(X)
+    cX = kmeans.cluster_centers_
+
+    # TODO: find a better way to choose neighborhood size
+    if n_features > 50:
+        k = 50
+    else:
+        k = 10
+
+    # In case some class has less elements than k
+    k_class = np.minimum(class_count, k)
+
+    # Construct index set with neighbors for every element of every class
+
+    idx_set = np.zeros((n_clusters, sum(k_class)), dtype=np.int)
+
+    start_finish_indices = np.hstack((0, k_class)).cumsum()
+
+    neigh = NearestNeighbors()
+
+    for c in range(n_class):
+        sel_c = np.where(y == labels[c])
+        kc = k_class[c]
+
+        # get k_class same class neighbours
+        neigh.fit(X=X[sel_c])
+
+        start, finish = start_finish_indices[c:c+2]
+        idx_set[:, start:finish] = np.take(sel_c, neigh.kneighbors(X=cX,
+                                           n_neighbors=kc,
+                                           return_distance=False))
+
+    # Compute basis for every cluster in first scale
+    basis = np.zeros((n_basis, n_features))
+    lda = LinearDiscriminantAnalysis()
+    for i, start in enumerate(range(0, num_eig * n_clusters, num_eig)):
+        lda.fit(X[idx_set[i, :]], y[idx_set[i, :]])
+        basis[start: start + num_eig, :] = normalize(lda.scalings_.T)
+
+    # second scale
+    k = 20
+
+    # In case some class has less elements than k
+    k_class = np.minimum(class_count, k)
+
+    # Construct index set with neighbors for every element of every class
+
+    idx_set = np.zeros((n_clusters, sum(k_class)), dtype=np.int)
+
+    start_finish_indices = np.hstack((0, k_class)).cumsum()
+
+    for c in range(n_class):
+      sel_c = np.where(y == labels[c])
+      kc = k_class[c]
+
+      # get k_class genuine neighbourss
+      neigh.fit(X=X[sel_c])
```

Comment:
Calls to `kneighbors` should be needed only once regardless of the number of scales, no? Using the max scale

Reply:
Great idea! This is a very good way to improve this!